### DISCUSSION TITLE
How to workaround dynamic indexing into pytrees?

### DISCUSSION THREAD

**Discussion by jatentaki** [jax-ml/jax#6107] [Q&A]
Hello,

I am looking to implement an RL setup of playing a card game. The game consists of multiple plies and every time the winner of the previous ply begins the next one. I want to have the agents be controlled by NNs, which carry their state as pytrees, so I run into the issue that in the game loop I need to dynamically index into a list of pytrees (to select and run the NN whose turn it is), which results in `TypeError: list indices must be integers or slices, not DynamicJaxprTracer`. Is there some standard workaround? I am interested in two scenarios:
1. when the networks have the same architecture (so we need to execute the same forward function, just with different sets of weights)
2. where the networks may be different (so we would need to select a _function_ to be ran, not just a pytree).

With 1., if I understand correctly, hand-rolling some awkward function to collapse the entire network state into a single contiguous jax array and storing agent states as a `[n_agents, n_weights_per_agent]` would allow me to dynamically select the weights in question. This looks very awkward, but at least makes it seem technically possible. Is 2. latter even possible? I'd like this thing to be possibly speedy, but I am aware that executing the various functions would probably have to be sequential.

Below is a minimal example to show what I mean (the easier case).

Thank you

```python
import jax
import jax.numpy as jnp
import jax.random as jr
import numpy as np

def play(rng, agent_state):
    '''
    in general, this decision would be obtained by running a neural network
    with parameters given in `agent_state`
    '''
    return jr.choice(rng, n_actions, shape=(), p=agent_state['p'])

def ply(rng, agent_states, order):
    '''
    Players take actions in order specified by the array of indexes `order`
    '''
    def body(carry, ix):
        in_rng, = carry
        out_rng, play_rng = jr.split(in_rng, 2)
        # The problem is here                \/
        action = play(play_rng, agent_states[ix])
        # we want to select agent states as
        # given by `order`, but because 
        # `agent_states` would in general be
        # a complex pytree of NN parameters,
        # we get
        # TypeError: list indices must be integers or slices, not DynamicJaxprTracer

        out_carry = out_rng,

        return out_carry, action

    _final_carry, actions = jax.lax.scan(
        body,
        (rng, ),
        order,
    )

    return actions

def game(rng, agent_states):
    keys = jr.split(rng, n_plies)

    # we play the first ply in order 0, 1, 2, 3, ...
    order = jnp.arange(len(agent_states))

    winners = -jnp.ones(n_plies)
    for i in range(n_plies):
        actions = ply(keys[i], agent_states, order)
        winner = actions.argmax()
        winners = winners.at[i].set(winner)

        # but later, it's always the previous ply's winner that
        # starts the next ply
        order = jnp.roll(order, winner)

    return winners

if __name__ == '__main__':
    n_players = 5
    n_actions = 7
    n_plies = 4
    rng_key = jr.PRNGKey(42)

    # define agent state as some action probabilities. In general, this would be
    # NN parameters
    agents = [
        {'p': np.random.uniform(size=(n_actions, ))} for _ in range(n_players)
    ]
    print(game(rng_key, agents))
```

**Comment by jakevdp** [comment]
I think the easiest way to make something like this compatible with JAX tracing is to change your data model from one that uses an array of structs to one that uses a struct of arrays. For example, your agents struct might look like this:
```python
agents = {'p': jnp.array(np.random.uniform(size=(n_players, n_actions)))}
```
Then you would have no issues indexing into the array with a tracer.

**Reply by jatentaki** [reply]
That's quite clever. I have so far used `haiku` for my neural nets, which represents state as nested dictionaries... I guess I can write some back-and-forth conversion code to (un)flatten them when needed. I am asking largely to avoid reinventing the wheel. Thanks!
